frontend/simple/js/event-queue.js

```diff
@@ -0,0 +1,50 @@
+import store from './state'
+var _queue = new WeakMap()
+var _waits = new WeakMap()
+// Event queue keeps track of event running in the background to prevent data corruption due to invalid state transitions
+// during event processing.
+class EventQueue {
+  constructor () {
+    _queue.set(this, [])
+    _waits.set(this, [])
+  }
+  // Events create a lock during their execution
+  lock () {
```

Comment:
FYI, going to look for a way to get rid of all lock/release stuff. Generally not a good idea to use locks (trust me, this comes from years of programming using locks)